pkg/controllers/monitoring/monitoring-stack/controller.go

```diff
@@ -145,6 +145,7 @@ func (rm resourceManager) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 			return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
 		}
 		if err != nil {
+			logger.Error(err, "failed to reconcile")
```

Comment:
IIUC, the default behaviour of the controller-runtime reconcile loop is to log any error is `Reconcile` returns it. Do you not see these errors in logs?

Reply:
I did not see any errors in the logs without this code.

My repro is like this:

- Deploy OCP 4.8 w/ Cluster Monitoring Operator
- Deploy ObO
- Create a MonitoringStack with `enableRemoteWriteReceiver: true`
- Observe that no Prometheus is created
- Check logs on ObO[1] and the ObO Prometheus Operator - they give no indication of a problem
- Wait a long time (over the weekend) and never see any log messages or Prometheus appear
- Apply this patch and redeploy
- Observe that there is an error in reconciling [2]

[1] It just repeats this over and over:
```
2022-09-06T16:14:23Z    INFO    observability-operator  Reconciling monitoring stack    {"stack": "observability-operator/sample-monitoring-stack"}
```
[2] New, helpful error message:
```
2022-09-06T16:54:59Z    ERROR   observability-operator  failed to reconcile      {"stack": "observability-operator/sample-monitoring-stack", "error": "failed to create typed patch object (observability-operator/sample-monitoring-stack; monitoring.coreos.com/v1, Kind=Prometheus): .spec.enableRemoteWriteReceiver: field not declared in schema"}
```

I'm also somewhat surprised that the framework did not report errors which appear to be returned (I didn't read further to figure out why); are they logged somewhere else? Please give my reproduction a test, or find another way to fail the reconciliation and let me know if I missed something, but I spent hours with no error message to follow up on and this helped me spot the problem (out of date CRD) in seconds.

Reply:
`updateStatus` being passed an `err` seems quite odd 🤔

Reply:
@csibbitt would you mind fixing the `updateStatus` to log the `recError` as an `INFO` ?

Reply:
I think that in this case, we want to return the reconciliation error instead of the status update error.
 
```
if err != nil {
    statusErr = rm.updateStatus(ctx, req, ms, err)
    if statusErr != nil {
        logger.Error(statusErr, "Failed to update status")
    }
    return err
}
```

Reply:
@sthaha It's possible that the error was reflected in the status of the MonitoringStack object and I simply missed it. I think logging is still a good idea.

I appreciate the effort going into making sure we do this the best way possible. I think Simon's latest suggestion makes the most sense if it results in displaying the error in the logs like we think it will. (though it might need to say `return ctrl.Result{}, err` I'll try this out locally when I'm done with my current experiments. :)

Reply:
Given that updateStatus can fail due to various reasons, (e.g. Get fails, updating status fails,) the error it returns should the reason for failure and log the `recError` which is passed to the function. 

WDYT @simonpasquier

Reply:
I think I agree with Simon. We should make sure the `Reconcile` function returns errors that are caused by the reconciliation logic.
I'd go as far as not `return rm.updateStatus(ctx, req, ms, err)` from `Reconcile` but to log these errors and return errors that pertain to reconciliation. The argument being that an error in updating the status is not tightly related to reconciliation of resources.

Reply:
I see your point but consider 2 places where updateStatus is called. 
```
r.updateStatus(ctx, req, ms, err) // when there is a know reconcile error
r.updateStatus(ctx, req, ms, nil) // when everything ran fine
```
In the first case if we returned the err, then operator will try and reconcile until everything runs fine (or for every) as opposed to current behaviour where operator logs failure in the status and only retries if the status couldn't be updated. 

Now, in the second case .. say there  was an issue updating the status, should the operator retry and update the status again? If we returned the `recError -> nil`, then operator won't retry, will it? 

How about we return recError if not nil else the error from statusUpdate?

Reply:
My initial idea was not to change the happy path (e.g. return error from updateStatus if any) but I feel like it makes the Reconcile() method less readable (sometimes updateStatus error is propagated to the caller, sometimes it's not).
Since `updateStatus` already returns a `crtl.Result` value, what if it didn't return any additional error but rather log errors caught when getting/updating the status and a "retry later" result?

```
func (rm resourceManager) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
        ...
        for _, reconciler := range reconcilers {
                err := reconciler.Reconcile(ctx, rm.k8sClient, rm.scheme)
                // handle create / update errors that can happen due to a stale cache by
                // retrying after some time.
                if errors.IsAlreadyExists(err) || errors.IsConflict(err) {
                        logger.V(3).Info("skipping reconcile error", "err", err)
                        return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
                }
                if err != nil {
                        return rm.updateStatus(ctx, req, ms, err), err
                }
        }

        return rm.updateStatus(ctx, req, ms, nil), nil
}

func (rm resourceManager) updateStatus(ctx context.Context, req ctrl.Request, ms *stack.MonitoringStack, recError error) ctrl.Result {
        ...
        err = rm.k8sClient.Status().Update(ctx, ms)
        if err != nil {
                logger.Info(statusErr, "Failed to update status")
                return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
        }
        return ctrl.Result{}, nil
}
```

Reply:
I like that @simonpasquier 🤗

Reply:
I'll finish this up based on Simon's recommendation. Thanks all.